### PR TITLE
3601 - [blackbox] sources should be persisted

### PR DESCRIPTION
**Pull request by Fischer-L**
Associated Issue: #3601

### Summary of Changes
Introduce "devtools.debugger.tabsBlackBoxed" pref to save the blackboxed sources and restore blackboxed states while receiving new sources.

### Screenshots/Videos 
ljp7tknlls

### CODE REVIEW

File: src/actions/sources/newSources.js [hunk 1]
```diff
@@ -155,8 +170,11 @@ export function newSources(sources: Source[]) {
       dispatch(checkPendingBreakpoints(source.id));
     }
 
-    return Promise.all(
+    await Promise.all(
       filteredSources.map(source => dispatch(loadSourceMap(source)))
     );
+    // We would like to restore the blackboxed state
+    // after loading all other states to make sure the correctness.
+    dispatch(resotoreBlackBoxedSources(filteredSources));
```
Comment (nyrosmith) [inline src/actions/sources/newSources.js]: Small typo: `resotoreBlackBoxedSources`

File: src/actions/sources/newSources.js [hunk 2]
```diff
@@ -124,6 +125,20 @@ function checkPendingBreakpoints(sourceId: string) {
   };
 }
 
+function resotoreBlackBoxedSources(sources: Source[]) {
```
Comment (nyrosmith) [inline src/actions/sources/newSources.js]: Small typo: `resotoreBlackBoxedSources`

Comment (Fischer-L) [reply src/actions/sources/newSources.js]: Thanks for spotting that

**Comment by jasonLaster** [comment]
Hey @bomsy @nyrosmith could be a good one to review

File: src/reducers/sources.js [hunk 3]
```diff
@@ -218,6 +221,27 @@ function updateTabList(state: OuterState, url: ?string, tabIndex?: number) {
...
+  }
+  prefs.tabsBlackBoxed = tabs;
+}
+
+export function getBlackBoxList() {
+  let tabs = prefs.tabsBlackBoxed;
+  if (!tabs) {
+    tabs = [];
+  }
+  return tabs;
```
Comment (darkwing) [inline src/reducers/sources.js]: This function could just be `return prefs.tabsBlackBoxed || [];`

File: src/actions/sources/newSources.js [hunk 4]
```diff
@@ -124,6 +125,20 @@ function checkPendingBreakpoints(sourceId: string) {
   };
 }
 
+function restoreBlackBoxedSources(sources: Source[]) {
+  return async ({ dispatch }: ThunkArgs) => {
+    const tabs = getBlackBoxList();
+    if (tabs.length <= 0) {
```
Comment (darkwing) [inline src/actions/sources/newSources.js]: This can probably just be `==`

Comment (Fischer-L) [reply src/actions/sources/newSources.js]: Updated, thanks

File: src/reducers/sources.js (hunk 3 above)
Comment (Fischer-L) [reply src/reducers/sources.js]: Updated, thanks

**Comment by jasonLaster** [comment]
I skimmed this and it looks good. I'll leave it to @nyrosmith and @darkwing to do the QA and review.

Thanks for the help!

**Pull request merged**